src/tpm2_packet.c

```diff
@@ -286,23 +286,83 @@ void TPM2_Packet_AppendAuthCmd(TPM2_Packet* packet, TPMS_AUTH_COMMAND* authCmd)
+ * otherwise adjusted based on the command information provided.
+ */
+int TPM2_GetCmdAuthCount(TPM2_CTX* ctx, CmdInfo_t* info)
+{
+    int authSessCount = 0, sessionCount;
+    TPMI_SH_AUTH_SESSION sessionHandle;
+    TPMA_SESSION sessionAttributes;
+    unsigned char flags = 0xFF;
+
+    if (info != NULL)
+        flags = info->flags;
+
+    /* The auth sessions must be first in the list */
+    for (sessionCount = 0; sessionCount < MAX_SESSION_NUM; sessionCount++) {
+        int authReq = 0;
+        sessionHandle = ctx->session[sessionCount].sessionHandle;
+        sessionAttributes = ctx->session[sessionCount].sessionAttributes;
+
+        if (info != NULL &&
+            ((sessionCount == 0 && (flags &
+                (CMD_FLAG_AUTH_USER1 |
+                 CMD_FLAG_AUTH_ADMIN |
+                 CMD_FLAG_AUTH_DUP))) ||
+             (sessionCount == 1 && (flags &
+                (CMD_FLAG_AUTH_USER2))))) {
+            authReq = 1;
+        }
 
-    TPM2_Packet_MarkU32(packet, &tmpSz);
-    for (i=0; i<authCount; i++) {
-        /* Note: Casting a TPM2_AUTH_SESSION to TPMS_AUTH_COMMAND here,
-            this is allowed because top of structure matches */
-        TPM2_Packet_AppendAuthCmd(packet, (TPMS_AUTH_COMMAND*)&ctx->session[i]);
+        /* Only a password auth if command user auth set */
+        if (sessionHandle == TPM_RS_PW && authReq) {
+            authSessCount++;
+        }
+
+        /* Only an HMAC session with encrypt, decrypt or audit set */
+        else if (authSessCount > 0 && TPM2_IS_HMAC_SESSION(sessionHandle)) {
+            if (((sessionAttributes & TPMA_SESSION_decrypt) && (flags &
+                    (CMD_FLAG_ENC2 | CMD_FLAG_ENC4))) ||
+                ((sessionAttributes & TPMA_SESSION_encrypt) && (flags &
+                    (CMD_FLAG_DEC2 | CMD_FLAG_DEC4))) ||
+                 (sessionAttributes & TPMA_SESSION_audit))
+                authSessCount++;
+        }
+    #ifdef DEBUG_WOLFTPM
+        else if (authReq) {
+            printf("Warning: Command requires auth at index %d!\n",
+                sessionCount);
+        }
+    #endif
     }
-    /* based on position difference places calculated size at marked U32 above */
-    TPM2_Packet_PlaceU32(packet, tmpSz);
+    return authSessCount;
+}
+
+TPM_ST TPM2_Packet_AppendAuth(TPM2_Packet* packet, TPM2_CTX* ctx, CmdInfo_t* info)
+{
+    TPM_ST st = TPM_ST_NO_SESSIONS;
 
-    return authCount;
+    if (ctx == NULL || info == NULL)
+        return BAD_FUNC_ARG;
+    if (ctx->session == NULL)
+        return st;
+
+    info->authCnt = TPM2_GetCmdAuthCount(ctx, info);
+    if (info->authCnt > 0) {
+        int i, tmpSz = 0;
+        TPM2_Packet_MarkU32(packet, &tmpSz);
+        for (i=0; i<info->authCnt; i++) {
+            /* Note: Casting a TPM2_AUTH_SESSION to TPMS_AUTH_COMMAND here,
+             * this is allowed because top of structure matches */
+            TPM2_Packet_AppendAuthCmd(packet, (TPMS_AUTH_COMMAND*)&ctx->session[i]);
```

Comment:
Is it possible for the indices to be mismatched here? For example suppose TPM2_GetCmdAuthCount incremented authSessCount for indices 0 and 2 with a count of 2. Wouldn't this use indices 0 and 1 instead of 0 and 2?